calculator.go

```diff
@@ -1,13 +1,122 @@
 // Package calculator provides a library for simple calculations in Go.
 package calculator
 
+import (
+	"errors"
+	"fmt"
+	"math"
+	"regexp"
+	"strconv"
+)
+
 // Add takes two numbers and returns the result of adding them together.
-func Add(a, b float64) float64 {
-return a + b
+func Add(a float64, b ...float64) float64 {
+	r := a + b[0]
+
+	// If b contains more than 1 element,
+	// call ourself with the current result and those remaining elements.
+	if len(b[1:]) > 0 {
+		r = Add(r, b[1:]...)
+	}
+	return r
 }
 
 // Subtract takes two numbers and returns the result of subtracting the second
 // from the first.
-func Subtract(a, b float64) float64 {
-	return b - a
+func Subtract(a float64, b ...float64) float64 {
+	r := a - b[0]
+
+	// If b contains more than 1 element,
+	// call ourself with the current result and those remaining elements.
+	if len(b[1:]) > 0 {
+		r = Subtract(r, b[1:]...)
+	}
+	return r
+}
+
+// Multiply takes two numbers and returns the result of multiplying them together.
+func Multiply(a float64, b ...float64) float64 {
+	r := a * b[0]
+
+	// If b contains more than 1 element,
+	// call ourself with the current result and those remaining elements.
+	if len(b[1:]) > 0 {
+		r = Multiply(r, b[1:]...)
+	}
+	return r
+}
+
+// Divide takes two numbers and returns the result of dividing the second
+// from the first.
+func Divide(a float64, b ...float64) (error, float64) {
+	var err error // potential error from recursive call to Divide()
+
+	if b[0] == 0 {
+		return errors.New(fmt.Sprintf("divide-by-zero for Divide(%f, %f)", a, b[0])), 0.0
+	}
+
+	r := a / b[0]
+
+	// If b contains more than 1 element,
+	// call ourself with the current result and those remaining elements.
+	if len(b[1:]) > 0 {
+		err, r = Divide(r, b[1:]...)
+		if err != nil {
+			return err, r
+		}
+	}
+	return nil, r
+}
+
+// Sqrt returns the square root of a number
+func Sqrt(a float64) (error, float64) {
+	if a == 0 {
+		return errors.New(fmt.Sprintf("cannot-get-square-root-of-a-negative-number for Sqrt(%f)", a)), 0.0
+	}
+
+	return nil, math.Sqrt(a)
+}
+
+func Expression(e string) (error, float64) {
+	// This regular expression matches:
+	//   (\d+\.?\d*) - one or more digits, optionally with a decimal and more digits
+	//   \s* - 0 or more space
+	//   ([\+\-\*\/]) - one of the operators: + - * or /
+	//   \s* - 0 or more space
+	//   (\d+\.?\d*) - one or more digits, optionally with a decimal and more digits
+	re := regexp.MustCompile(`^(\d+\.?\d*)\s*([\+\-\*\/])\s*(\d+\.?\d*)`)
+	fields := re.FindStringSubmatch(e)
```

Comment:
Here, we could use something like `fmt.Sscanf`, because the input format is pretty simple, without any complications that would require a regular expression.